### PR TITLE
Perform physical copy in gateway

### PR DESCRIPTION
**Pull request by guy-har**

### CODE REVIEW

File: pkg/gateway/operations/putobject.go [hunk 1]
```diff
@@ -74,11 +73,22 @@ func handleCopy(w http.ResponseWriter, req *http.Request, o *PathOperation, copy
 		return // operation already failed
 	}
 	ent.CreationDate = time.Now()
-	ent.Path = o.Path
-	err := o.Catalog.CreateEntry(req.Context(), o.Repository.Name, o.Reference, *ent)
+	address, err := o.Catalog.GenerateName()
 	if err != nil {
-		o.Log(req).WithError(err).Error("could not write copy destination")
-		_ = o.EncodeError(w, req, errors.Codes.ToAPIErr(errors.ErrInvalidCopyDest))
+		_ = o.EncodeError(w, req, errors.Codes.ToAPIErr(errors.ErrInternalError))
```
Comment (arielshaqed) [inline pkg/gateway/operations/putobject.go:78]: Can we have a more detailed error than "internal error"?  Also we will need some way to debug this, and we either have no log or we lost some detail in the log.

File: pkg/gateway/operations/putobject.go [hunk 2]
```diff
@@ -74,11 +73,22 @@ func handleCopy(w http.ResponseWriter, req *http.Request, o *PathOperation, copy
...
+	destinationObjectPointer := block.ObjectPointer{StorageNamespace: o.Repository.StorageNamespace, Identifier: address}
+
+	err = o.BlockStore.Copy(req.Context(), sourceObjectPointer, destinationObjectPointer)
+	if err != nil {
+		_ = o.EncodeError(w, req, errors.Codes.ToAPIErr(errors.ErrInternalError))
+		return
+	}
+	err = o.finishUpload(req, ent.Checksum, address, ent.Size, true)
+	if err != nil {
+		_ = o.EncodeError(w, req, errors.Codes.ToAPIErr(errors.ErrInternalError))
```
Comment (arielshaqed) [inline pkg/gateway/operations/putobject.go:91]: I don't understand: what happened to the previous `InvalidCopyDest` error return?

File: pkg/gateway/operations/putobject.go [hunk 3]
```diff
@@ -217,7 +227,13 @@ func handlePut(w http.ResponseWriter, req *http.Request, o *PathOperation) {
 	o.Incr("put_object")
 	storageClass := StorageClassFromHeader(req.Header)
 	opts := block.PutOpts{StorageClass: storageClass}
-	blob, err := upload.WriteBlob(req.Context(), o.BlockStore, o.Repository.StorageNamespace, req.Body, req.ContentLength, opts)
+	physicalAddress, err := o.Catalog.GenerateName()
+	if err != nil {
+		o.Log(req).WithError(err).Error("failed generating a name for physical address")
```
Comment (arielshaqed) [inline pkg/gateway/operations/putobject.go:232]:
```suggestion
		o.Log(req).WithError(err).Error("generate physical address")
```

**Review by arielshaqed — COMMENTED** [review]
Code itself looks cool, but still not sure about implementing this requirement; we may have some additional data soon about Spark needing renames, which this will make considerably more expensive.

**Review by johnnyaug — APPROVED** [review]
LGTM

**Comment by guy-har** [comment]
We decided not to perform physical copy for now, Garbage collections and lifecycle should handle deduplications

**Pull request closed without merging**